src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs

```diff
@@ -0,0 +1,280 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Net.Http;
+using System.Security.Cryptography;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Azure.WebJobs.Host.Executors;
+using Microsoft.Azure.WebJobs.Script.Description;
+using Microsoft.Azure.WebJobs.Script.Rpc;
+using Microsoft.Azure.WebJobs.Script.WebHost.Extensions;
+using Microsoft.Azure.WebJobs.Script.WebHost.Security;
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using Microsoft.WindowsAzure.Storage;
+using Microsoft.WindowsAzure.Storage.Blob;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.Azure.WebJobs.Script.WebHost.Management
+{
+    public class FunctionsSyncManager : IFunctionsSyncManager
+    {
+        private const string HubName = "HubName";
+        private const string TaskHubName = "taskHubName";
+        private const string Connection = "connection";
+        private const string DurableTaskStorageConnectionName = "azureStorageConnectionStringName";
+        private const string DurableTask = "durableTask";
+
+        private readonly IOptionsMonitor<ScriptApplicationHostOptions> _applicationHostOptions;
+        private readonly ILogger _logger;
+        private readonly IEnumerable<WorkerConfig> _workerConfigs;
+        private readonly HttpClient _httpClient;
+        private readonly ISecretManagerProvider _secretManagerProvider;
+        private readonly IConfiguration _configuration;
+        private readonly IHostIdProvider _hostIdProvider;
+
+        private CloudBlockBlob _hashBlob;
+
+        public FunctionsSyncManager(IConfiguration configuration, IHostIdProvider hostIdProvider, IOptionsMonitor<ScriptApplicationHostOptions> applicationHostOptions, IOptions<LanguageWorkerOptions> languageWorkerOptions, ILoggerFactory loggerFactory, HttpClient httpClient, ISecretManagerProvider secretManagerProvider)
+        {
+            _applicationHostOptions = applicationHostOptions;
+            _logger = loggerFactory?.CreateLogger(ScriptConstants.LogCategoryHostGeneral);
+            _workerConfigs = languageWorkerOptions.Value.WorkerConfigs;
+            _httpClient = httpClient;
+            _secretManagerProvider = secretManagerProvider;
+            _configuration = configuration;
+            _hostIdProvider = hostIdProvider;
+        }
+
+        private CloudBlockBlob HashBlob
```

Comment:
At some point we will need an implementation of this that works without azure storage (any persistent storage is sufficient) - for example, when running in on-prem Kubernetes. As a result, you'll probably need to take the storage specific parts of this implementation and put them behind an interface.

 I wouldn't block this review on it (the sooner we get this in the better, decoupling from storage can happen later), just mentioning this if you have to factor the code further as part of other review feedback.